Mathlib/AlgebraicGeometry/EllipticCurve/Affine/Basic.lean

```diff
@@ -189,99 +189,103 @@ noncomputable def polynomialY : R[X][Y] :=
   eval_simp
-  rw [← add_assoc]
+  rw [add_assoc]
 
 @[simp]
 lemma evalEval_polynomialY_zero : W.polynomialY.evalEval 0 0 = W.a₃ := by
-  simp only [evalEval_polynomialY, zero_add, mul_zero]
+  simp_rw [evalEval_polynomialY, mul_zero, zero_add]
 
 variable (W) in
 /-- The proposition that an affine point `(x, y)` on a Weierstrass curve `W` is nonsingular.
 
-In other words, either `W_X(x, y) ≠ 0` or `W_Y(x, y) ≠ 0`.
+In other words, the ideal spanned by `W_X(x, y)` and `W_Y(x, y)` is the entire ring.
 
-Note that this definition is only mathematically accurate for fields. -/
--- TODO: generalise this definition to be mathematically accurate for a larger class of rings.
+Note that this definition is only mathematically accurate for certain rings whose Picard group has
+trivial 12-torsion, such as a field or a PID. -/
 def Nonsingular (x y : R) : Prop :=
-  W.Equation x y ∧ (W.polynomialX.evalEval x y ≠ 0 ∨ W.polynomialY.evalEval x y ≠ 0)
-
-lemma nonsingular_iff' (x y : R) : W.Nonsingular x y ↔ W.Equation x y ∧
-    (W.a₁ * y - (3 * x ^ 2 + 2 * W.a₂ * x + W.a₄) ≠ 0 ∨ 2 * y + W.a₁ * x + W.a₃ ≠ 0) := by
-  rw [Nonsingular, equation_iff', evalEval_polynomialX, evalEval_polynomialY]
+  W.Equation x y ∧ Ideal.span {W.polynomialX.evalEval x y, W.polynomialY.evalEval x y} = ⊤
 
 lemma nonsingular_iff (x y : R) : W.Nonsingular x y ↔ W.Equation x y ∧
-    (W.a₁ * y ≠ 3 * x ^ 2 + 2 * W.a₂ * x + W.a₄ ∨ y ≠ -y - W.a₁ * x - W.a₃) := by
-  rw [nonsingular_iff', sub_ne_zero, ← sub_ne_zero (a := y)]
-  congr! 3
-  ring1
+    Ideal.span {W.a₁ * y - (3 * x ^ 2 + 2 * W.a₂ * x + W.a₄), 2 * y + W.a₁ * x + W.a₃} = ⊤ := by
+  rw [Nonsingular, evalEval_polynomialX, evalEval_polynomialY]
+
+@[deprecated (since := "2025-05-26")] alias nonsingular_iff' := nonsingular_iff
 
 @[simp]
-lemma nonsingular_zero : W.Nonsingular 0 0 ↔ W.a₆ = 0 ∧ (W.a₃ ≠ 0 ∨ W.a₄ ≠ 0) := by
-  rw [Nonsingular, equation_zero, evalEval_polynomialX_zero, neg_ne_zero, evalEval_polynomialY_zero,
-    or_comm]
+lemma nonsingular_zero : W.Nonsingular 0 0 ↔ W.a₆ = 0 ∧ Ideal.span {W.a₃, W.a₄} = ⊤ := by
+  rw [Nonsingular, equation_zero, evalEval_polynomialX_zero, Ideal.span_pair_neg_left,
+    evalEval_polynomialY_zero, Ideal.span_pair_comm]
 
 lemma nonsingular_iff_variableChange (x y : R) :
     W.Nonsingular x y ↔ (VariableChange.mk 1 x 0 y • W).toAffine.Nonsingular 0 0 := by
-  rw [nonsingular_iff', equation_iff_variableChange, equation_zero, ← neg_ne_zero, or_comm,
-    nonsingular_zero, variableChange_a₃, variableChange_a₄, inv_one, Units.val_one]
-  simp only [variableChange_def]
-  congr! 3 <;> ring1
+  rw [nonsingular_iff, equation_iff_variableChange, equation_zero, ← neg_sub,
+    Ideal.span_pair_neg_left, Ideal.span_pair_comm, nonsingular_zero, variableChange_a₃,
+    variableChange_a₄, inv_one, Units.val_one]
+  ring_nf
 
-private lemma equation_zero_iff_nonsingular_zero_of_Δ_ne_zero (hΔ : W.Δ ≠ 0) :
+private lemma equation_zero_iff_nonsingular_zero_of_isUnit_Δ (hΔ : IsUnit W.Δ) :
```

Comment:
it might be useful (to the computer scientists) to extract a lemma separately saying that the polynomial Δ is in the ideal generated by a3, a4, and a6.